Source/JavaScriptCore/runtime/JSONObject.cpp

```diff
@@ -627,6 +637,517 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+                recordFailure("string character needs escaping"_s);
+                return;
+            }
+            m_buffer[m_length + 1 + i] = character;
+        }
+        m_buffer[m_length + 1 + stringLength] = '"';
+        m_length += 1 + stringLength + 1;
+        return;
+    }
+
+    case ObjectType:
+    case FinalObjectType: {
+        auto& object = *asObject(&cell);
+        if (object.isCallable()) {
+            recordFailure("callable object"_s);
+            return;
+        }
+        auto* ancestorClass = object.classInfo();
+        do {
+            if (ancestorClass == NumberObject::info()) {
+                recordFailure("wrapped number"_s);
+                return;
+            }
+            if (ancestorClass == StringObject::info()) {
+                recordFailure("wrapped string"_s);
+                return;
+            }
+            if (ancestorClass == BooleanObject::info()) {
+                recordFailure("wrapped boolean"_s);
+                return;
+            }
+            if (ancestorClass == BigIntObject::info()) {
+                recordFailure("wrapped big int"_s);
+                return;
+            }
+        } while ((ancestorClass = ancestorClass->parentClass));
+        auto& structure = *object.structure();
+        if (object.structureID() != structure.id()) {
+            recordFailure("structure ID mismatch"_s);
+            return;
+        }
+        if (structure.hasPolyProto()) {
+            recordFailure("hasPolyProto"_s);
+            return;
+        }
+        if (structure.storedPrototype() != m_globalObject.objectPrototype()) {
+            recordFailure("non-standard object prototype"_s);
+            return;
+        }
+        if (!m_checkedObjectPrototype) {
+            if (mayHaveToJSON(*m_globalObject.objectPrototype())) {
+                recordFailure("object prototype may have toJSON"_s);
+                return;
+            }
+            m_checkedObjectPrototype = true;
+        }
+        if (!remainingCapacity()) {
+            recordBufferFull();
+            return;
+        }
+        m_buffer[m_length++] = '{';
+        if (canPerformFastPropertyEnumerationForJSONStringify(&structure)) {
+            structure.forEachProperty(m_vm, [&](const PropertyTableEntry& entry) -> bool {
+                if (entry.attributes() & PropertyAttribute::DontEnum)
+                    return true;
+                auto& name = *entry.key();
+                if (name.isSymbol()) {
+                    recordFailure("symbol"_s);
+                    return false;
+                }
+                if (!name.is8Bit()) {
+                    recordFailure("16-bit property name"_s);
+                    return false;
+                }
+                bool needComma = m_buffer[m_length - 1] != '{';
+                unsigned nameLength = name.length();
+                if (remainingCapacity() < needComma + 1 + nameLength + 2) {
```

Comment:
BTW, String.length()'s maximum value is INT32_MAX, so, this never overflows. (so this is fine).